midend/hsIndexSimplify.cpp

```diff
@@ -0,0 +1,201 @@
+        arrayIndex->right->is<IR::PathExpression>())) {
+        // Generate new local variable if needed.
+        std::ostringstream ostr;
+        ostr << arrayIndex->right;
+        cstring indexString = ostr.str();
+        if (arrayIndex->right->is<IR::PathExpression>()) {
+            newVariable = arrayIndex->right->to<IR::PathExpression>();
+        } else if (generatedVariables->count(indexString) == 0) {
+            // Generate new temporary variable.
+            auto type = typeMap->getTypeType(arrayIndex->right->type, true);
+            auto name = refMap->newName("hsiVar");
+            auto decl = new IR::Declaration_Variable(name, type);
+            locals->push_back(decl);
+            typeMap->setType(decl, type);
+            newVariable = new IR::PathExpression(arrayIndex->srcInfo, type, new IR::Path(name));
+            generatedVariables->emplace(indexString, newVariable);
+        } else {
+            newVariable = generatedVariables->at(indexString);
+        }
+    }
+}
+
+const IR::Node* HSIndexTransform::postorder(IR::ArrayIndex* curArrayIndex) {
+    if (hsIndexFinder.arrayIndex != nullptr &&
+        curArrayIndex->equiv(*hsIndexFinder.arrayIndex)) {
+        // Translating current array index.
+        auto* newArrayIndex = hsIndexFinder.arrayIndex->clone();
+        newArrayIndex->right = new IR::Constant(newArrayIndex->right->type, index);
+        return newArrayIndex;
+    }
+    return curArrayIndex;
+}
+
+IR::Node* HSIndexSimplifier::eliminateArrayIndexes(HSIndexFinder& aiFinder,
+                                                   IR::Statement* statement,
+                                                   const IR::Expression* expr) {
+    if (aiFinder.arrayIndex == nullptr) {
+        return statement;
+    }
+    const IR::Statement* elseBody = nullptr;
+    auto* curIf = statement->to<IR::IfStatement>();
+    if (curIf != nullptr) {
+        elseBody = curIf->ifFalse;
+    }
+    IR::IndexedVector<IR::StatOrDecl> newComponents;
+    if (aiFinder.newVariable != nullptr) {
+        if (!aiFinder.newVariable->equiv(*aiFinder.arrayIndex->right)) {
+            newComponents.push_back(new IR::AssignmentStatement(aiFinder.arrayIndex->srcInfo,
+            aiFinder.newVariable, aiFinder.arrayIndex->right));
+        }
+    }
+    IR::IfStatement* result = nullptr;
+    IR::IfStatement* curResult = nullptr;
+    IR::IfStatement* newIf = nullptr;
+    const auto* typeStack = aiFinder.arrayIndex->left->type->checkedTo<IR::Type_Stack>();
+    size_t sz = typeStack->getSize();
+    for (size_t i = 0; i < sz; i++) {
+        HSIndexTransform aiRewriter(aiFinder, i);
+        auto* newStatement = statement->apply(aiRewriter)->to<IR::Statement>();
+        auto* newIndex = new IR::Constant(aiFinder.arrayIndex->right->type, i);
+        auto* newCondition = new IR::Equ(aiFinder.newVariable, newIndex);
+        if (curIf != nullptr) {
+            newIf = newStatement->to<IR::IfStatement>()->clone();
```

Comment:
Make this a `checkedTo()` since this will fail with a nullptr access if newStatement is not an IfStatement.